.github/actions/verify-token/src/predicate.ts

```diff
@@ -0,0 +1,219 @@
+    env.GITHUB_REF || ""
+  );
+
+  const predicate: SLSAv1Predicate = {
+    buildDefinition: {
+      buildType: DELEGATOR_BUILD_TYPE,
+      externalParameters: {
+        // This is the v0.2 entryPoint.
+        workflowPath: { value: currentRun.path },
+        // We only use source here because the source contained the source
+        // repository and the build configuration.
+        source: {
+          artifact: {
+            uri: callerRepo,
+            digest: {
+              sha1: env.GITHUB_SHA || "",
+            },
+          },
+        },
+      },
+      systemParameters: {
+        // TODO(https://github.com/slsa-framework/slsa-github-generator/issues/1505):
+        // Add GitHub event payload.
+        GITHUB_EVENT_NAME: { value: String(env.GITHUB_EVENT_NAME) },
+        GITHUB_JOB: { value: String(env.GITHUB_JOB) },
+        GITHUB_REF: { value: String(env.GITHUB_REF) },
+        GITHUB_REF_TYPE: { value: String(env.GITHUB_REF_TYPE) },
+        GITHUB_REPOSITORY: { value: String(env.GITHUB_REPOSITORY) },
+        GITHUB_RUN_ATTEMPT: { value: String(env.GITHUB_RUN_ATTEMPT) },
+        GITHUB_RUN_ID: { value: String(env.GITHUB_RUN_ID) },
+        GITHUB_RUN_NUMBER: { value: String(env.GITHUB_RUN_NUMBER) },
+        GITHUB_SHA: { value: String(env.GITHUB_SHA) },
+        GITHUB_WORKFLOW: { value: String(env.GITHUB_WORKFLOW) },
+        GITHUB_ACTOR_ID: { value: String(currentRun.actor?.id) },
+        GITHUB_REPOSITORY_ID: { value: String(currentRun.repository.id) },
+        GITHUB_REPSITORY_OWNER_ID: {
+          value: String(currentRun.repository.owner.id),
+        },
+        GITHUB_WORKFLOW_REF: { value: String(env.GITHUB_WORKFLOW_REF) },
+        GITHUB_WORKFLOW_SHA: { value: String(env.GITHUB_WORKFLOW_SHA) },
+        IMAGE_OS: { value: String(env.ImageOS) },
+        IMAGE_VERSION: { value: String(env.ImageVersion) },
+        RUNNER_ARCH: { value: String(env.RUNNER_ARCH) },
+        RUNNER_NAME: { value: String(env.RUNNER_NAME) },
+        RUNNER_OS: { value: String(env.RUNNER_OS) },
+      },
+    },
+    runDetails: {
+      // TODO(https://github.com/slsa-framework/slsa-github-generator/issues/1504):
+      // Record raw token inputs (tool action inputs).
+      builder: {
+        id: toolURI,
+      },
+      metadata: {
+        invocationId: `https://github.com/${currentRun.repository.full_name}/actions/runs/${currentRun.id}/attempts/${currentRun.run_attempt}`,
+      },
+    },
+  };
+
+  if (env.GITHUB_EVENT_NAME === "workflow_dispatch") {
+    if (env.GITHUB_EVENT_PATH !== undefined) {
+      const ghEvent: WorkflowDispatchEvent = JSON.parse(
+        fs.readFileSync(env.GITHUB_EVENT_PATH).toString()
+      );
+
+      for (const input in ghEvent.inputs) {
+        // The invocation parameters belong here and are the top-level GitHub
+        // workflow inputs.
+        predicate.buildDefinition.externalParameters[`input_${input}`] = {
+          value: String(ghEvent.inputs[input]),
```

Comment:
> We may also be able to assume that inputs[input] is a string and just do ghEvent.inputs[input] || ""

Even if not a string, it would still work to do `String(ghEvent.inputs[input] || "")`. We always serialize it into a string, but it could be a number and `String(number)` will take care of that.